### DISCUSSION TITLE
Visual studio error LNK2019 unresolved external symbol

### DISCUSSION THREAD

**Discussion by CES1919** [ihedvall/mdflib#75] [Q&A] [answered]
I’m a rookie about C++ programming , so I tried for a long time but couldn't figure out what was causing the errors. 
      when compiled a mdfd.lib ，I wish to call this .lib to read mf4 files. but the visual studio 2019 show errors as follows：  

>       Error	LNK2019	unresolved external symbol XML_ErrorString referenced in function "public: virtual bool __cdecl mdf::ExpatXml::ParseFile(void)" (?ParseFile@ExpatXml@mdf@@UEAA_NXZ)	ParserMf4	C:\WORK\script\C++\ParserMf4\mdfd.lib(expatxml.obj)	1	
> Error	LNK2019	unresolved external symbol XML_ParserFree referenced in function "public: __cdecl `anonymous namespace'::ExpatParser::~ExpatParser(void)" (??1ExpatParser@?A0x0c9854eb@@QEAA@XZ)	ParserMf4	C:\WORK\script\C++\ParserMf4\mdfd.lib(expatxml.obj)	1	
> Error	LNK2019	unresolved external symbol XML_GetCurrentColumnNumber referenced in function "public: virtual bool __cdecl mdf::ExpatXml::ParseFile(void)" (?ParseFile@ExpatXml@mdf@@UEAA_NXZ)	ParserMf4	C:\WORK\script\C++\ParserMf4\mdfd.lib(expatxml.obj)	1	
> Error	LNK2019	unresolved external symbol XML_GetCurrentLineNumber referenced in function "public: virtual bool __cdecl mdf::ExpatXml::ParseFile(void)" (?ParseFile@ExpatXml@mdf@@UEAA_NXZ)	ParserMf4	C:\WORK\script\C++\ParserMf4\mdfd.lib(expatxml.obj)	1	
> Error	LNK2019	unresolved external symbol XML_GetErrorCode referenced in function "public: virtual bool __cdecl mdf::ExpatXml::ParseFile(void)" (?ParseFile@ExpatXml@mdf@@UEAA_NXZ)	ParserMf4	C:\WORK\script\C++\ParserMf4\mdfd.lib(expatxml.obj)	1	
     
  These errors seem to be caused by the library not being linked, but I checked that the lib is indeed linked. so am I not linked something somewhere, or am I linked something wrong?

  The steps I did are as follows：

**1. Download mdflib-main.zip and use cmake generate visual studio project  for mdflib** 
![image](https://github.com/ihedvall/mdflib/assets/85708292/30ca774f-6bb2-48cc-8917-6a9d3be69d18)
![image](https://github.com/ihedvall/mdflib/assets/85708292/1e5760a1-a6cc-4ae6-a73b-cfd24f066bd0)
![image](https://github.com/ihedvall/mdflib/assets/85708292/7e99a178-3dc0-4c37-b14e-ad15f0f1d45f)
![image](https://github.com/ihedvall/mdflib/assets/85708292/898401c0-6367-4161-89d1-ddc4f55795a7)
![image](https://github.com/ihedvall/mdflib/assets/85708292/5c06c830-66ac-4191-86b1-18be19a38270)
![image](https://github.com/ihedvall/mdflib/assets/85708292/76a4ffbd-b71f-4ad0-8aa0-a374ab9af5dc)

**2.build mdfd.lib  by project** 
![image](https://github.com/ihedvall/mdflib/assets/85708292/c09418a2-271b-4546-81b8-f34c1232e261)
![image](https://github.com/ihedvall/mdflib/assets/85708292/988a87ac-1660-4418-ac77-9cadeb8c5ab7)

**3.Creat a new project to callmdfd.lib, but show errors**
![image](https://github.com/ihedvall/mdflib/assets/85708292/69c7b2cd-e00e-4dc5-b273-8f5208647979)
![image](https://github.com/ihedvall/mdflib/assets/85708292/40353a46-d56f-4e5e-bb8e-115154638518)

Thanks for any answer！

**Comment by ihedvall** [comment, marked as answer]
Don't panic. 

The ParseMf4.exe project is missing both zlib and expat libraries. The C++ linkers are not that smart regarding this common error. The errors are generated when linking the executable, so you need to manually add these external libraries to your executable project.

I think it should be the static version (MT) version of the libraries you should use and yes, the MT and MD option in VC is just weird causing a lot of the above problem.

**Reply by CES1919** [reply]
Yes ,you are right !  Although I manually added zlib and expat libraries to my executable project. But these libraries are incorrect (I Stupidly  installed these libraries using the 32bit expat.exe and the old version windows zlib.exe). When I re-downloaded the source code for zlib and expat and then compiled and installed.  The error is solved.
  Thanks for your efficient answer!
